File: packages/b2c-tooling-sdk/src/auth/basic.ts
/*
 * Copyright (c) 2025, Salesforce, Inc.
 * SPDX-License-Identifier: Apache-2
 * For full license text, see the license.txt file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
 */
import type {AuthStrategy, FetchInit} from './types.js';
import {dispatchFetch} from './dispatch-fetch.js';
import {getLogger} from '../logging/logger.js';

/**
 * Basic authentication strategy for WebDAV operations.
 *
 * Encodes username and access key as Base64 for HTTP Basic auth.
 * Implements the {@link AuthStrategy} interface.
 *
 * @example
 * ```typescript
 * import { BasicAuthStrategy } from '@salesforce/b2c-tooling-sdk';
 *
 * const auth = new BasicAuthStrategy('username', 'access-key');
 * const response = await auth.fetch('https://webdav.example.com/path');
 * ```
 */
export class BasicAuthStrategy implements AuthStrategy {
  private encoded: string;

  /**
   * Creates a new BasicAuthStrategy instance for HTTP Basic authentication.
   *
   * @param user - The username for Basic authentication
   * @param pass - The password or access key for Basic authentication
   */
  constructor(user: string, pass: string) {
    this.encoded = Buffer.from(`${user}:${pass}`).toString('base64');

    const logger = getLogger();
    logger.debug({username: user}, `[Auth] Using Basic authentication for user: ${user}`);
  }

  async fetch(url: string, init: FetchInit = {}): Promise<Response> {
    const headers = new Headers(init.headers);
    headers.set('Authorization', `Basic ${this.encoded}`);
    // Pass through dispatcher for TLS/mTLS support (see dispatchFetch)
    return dispatchFetch(url, {...init, headers});
  }

  async getAuthorizationHeader(): Promise<string> {
    return `Basic ${this.encoded}`;
  }
}

File: python/src/b2c_tooling_sdk/auth/basic.py
# Copyright (c) 2025, Salesforce, Inc.
# SPDX-License-Identifier: Apache-2.0
# For full license text, see the LICENSE file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
"""HTTP Basic authentication strategy (used for WebDAV).

Mirrors ``src/auth/basic.ts``. Encodes username and access key as Base64 for the
``Authorization: Basic`` header.
"""

from __future__ import annotations

import base64
from typing import Any

import httpx

from b2c_tooling_sdk.auth.dispatch_fetch import dispatch_fetch
from b2c_tooling_sdk.logging import get_logger


class BasicAuthStrategy:
    """Basic authentication strategy.

    :example:

    .. code-block:: python

        from b2c_tooling_sdk.auth import BasicAuthStrategy

        auth = BasicAuthStrategy("username", "access-key")
        response = await auth.fetch("https://webdav.example.com/path")
    """

    def __init__(self, user: str, password: str) -> None:
        self._encoded = base64.b64encode(f"{user}:{password}".encode()).decode("ascii")
        get_logger("auth.basic").debug("[Auth] Using Basic authentication for user: %s", user)

    async def fetch(
        self,
        url: str,
        *,
        method: str = "GET",
        headers: dict[str, str] | None = None,
        content: Any = None,
        dispatcher: httpx.AsyncBaseTransport | None = None,
        **kwargs: Any,
    ) -> httpx.Response:
        """Perform a request with the ``Authorization: Basic`` header set."""
        request_headers = dict(headers or {})
        request_headers["Authorization"] = f"Basic {self._encoded}"
        return await dispatch_fetch(
            url, method=method, headers=request_headers, content=content, dispatcher=dispatcher, **kwargs
        )

    async def get_authorization_header(self) -> str:
        """Return the ``Authorization: Basic`` header value."""
        return f"Basic {self._encoded}"


__all__ = ["BasicAuthStrategy"]
